peps/pep-0734.rst

```diff
@@ -0,0 +1,697 @@
+automatically for the initial ("main") thread and for
+``threading.Thread`` objects.  From the C-API it is set (and cleared)
+by ``PyThreadState_Swap()`` and may be set by ``PyGILState_Ensure()``.
+Most of the C-API requires that there be a current thread state,
+either looked up implicitly or passed in as an argument.
+
+The relationship between threads and thread states in one-to-many.
+Each thread state is associated with at most a single thread and
+records its thread ID.  It is never used for another thread.
+In the other direction, however, a thread may have more than one thread
+state associated with it, though, again, only one may be current.
+
+When there's more than one thread state for a thread,
+``PyThreadState_Swap()`` is used to switch between them,
+with the requested thread state becoming the current one.
+Whatever was running in the thread using the old thread state
+is effectively paused until that thread state is swapped back in.
+
+Interpreter States
+------------------
+
+As noted earlier, there is some runtime state that multiple threads
+share.  Some of it is exposed by the ``sys`` module, though much is
+used internally and not exposed explicitly or only through the C-API.
+
+This state that thread states share is called the interpreter state
+(``PyInterpreterState``).  We'll sometimes refer to it here as just
+"interpreter", though that is also sometimes used to refer to the
+``python`` executable, to the Python implementation, and to the
+bytecode interpreter (i.e. ``exec()``/``eval()``).
+
+CPython has supported multiple interpreters in the same process (AKA
+"subinterpreters") since version 1.5 (1997).  The feature has been
+available via the :ref:`C API <python:sub-interpreter-support>`.
+
+Interpreters and Threads
+------------------------
+
+There's a one-to-many relationship between thread states and interpreter
+states.  A thread state belongs to a single interpreter and stores
+a pointer to it.  That thread state is never used for a different
+interpreter.  In the other direction, however, an interpreter may have
+more than thread state associated with it.  The interpreter is only
+considered active in threads where one of its thread states is current.
+
+Interpreters are created via the C-API using ``Py_NewInterpreter()``
+or ``Py_NewInterpreterFromConfig()``.  Both functions do the following:
+
+1. create a new interpreter state
+2. create a new thread state
+3. set the thread state as current
+4. initialize the interpreter state using that thread state
+5. return the thread state (still current)
+
+To make an existing interpreter active in the current thread,
+we first make sure it has a thread state for the thread.  Then
+we call ``PyThreadState_Swap()`` like normal using that thread state.
+If the thread state for another interpreter was already current then
+it gets swapped out like normal and execution of that interpreter in
+the thread is thus effectively paused until it is swapped back in.
```

Comment:
> Now I'm curious what happens if _no_ thread state was already current.

If there wasn't one then nothing's really that different.  The new interpreter get created and its new thread state gets set as current.

> Or if another thread state for the same interpreter was current.

That shouldn't be possible since we're just barely creating the new interpreter.  There wouldn't be any opportunity to create a second thread state for that interpreter.